server/channels/jobs/batch_report_worker.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+	"github.com/pkg/errors"
+)
+
+type BatchReportWorkerAppIFace interface {
+	SaveReportChunk(format string, prefix string, count int, reportData []model.ReportableObject) *model.AppError
+	CompileReportChunks(format string, prefix string, numberOfChunks int, headers []string) (string, *model.AppError)
+	SendReportToUser(userID string, filename string) *model.AppError
+}
+
+type BatchReportWorker[T BatchReportWorkerAppIFace] struct {
+	BatchWorker[T]
+	reportFormat string
+	headers      []string
+	getData      func(jobData model.StringMap, app T) ([]model.ReportableObject, model.StringMap, bool, error)
+}
+
+func MakeBatchReportWorker[T BatchReportWorkerAppIFace](
+	jobServer *JobServer,
+	store store.Store,
+	app T,
+	timeBetweenBatches time.Duration,
+	reportFormat string,
+	headers []string,
+	getData func(jobData model.StringMap, app T) ([]model.ReportableObject, model.StringMap, bool, error),
+) model.Worker {
+	worker := &BatchReportWorker[T]{
+		reportFormat: reportFormat,
+		headers:      headers,
+		getData:      getData,
+	}
+	worker.BatchWorker = BatchWorker[T]{
+		jobServer:          jobServer,
+		logger:             jobServer.Logger(),
+		store:              store,
+		app:                app,
+		stop:               make(chan struct{}),
+		stopped:            make(chan bool, 1),
+		jobs:               make(chan model.Job),
+		timeBetweenBatches: timeBetweenBatches,
+		doBatch:            worker.doBatch,
+	}
+	return worker
+}
+
+func (worker *BatchReportWorker[T]) doBatch(rctx *request.Context, job *model.Job) bool {
+	reportData, nextData, done, err := worker.getData(job.Data, worker.app)
+	if err != nil {
+		// TODO getData error
+		worker.logger.Error("Worker: Failed to do report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	} else if done {
+		if err = worker.complete(job); err != nil {
+			// TODO complete error
+			worker.logger.Error("Worker: Failed to do report batch. Exiting", mlog.Err(err))
+			worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		} else {
+			worker.logger.Info("Worker: Job is complete")
+			worker.setJobSuccess(worker.logger, job)
+		}
+
+		return true
+	}
+
+	err = worker.saveData(job, reportData)
+	if err != nil {
+		// TODO saveData error
+		worker.logger.Error("Worker: Failed to do report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	}
+
+	for k, v := range nextData {
+		job.Data[k] = v
+	}
+
+	// TODO add progress?
```

Comment:
This might be doable if it's not too inefficient to count the number of pages that we will have.